webui/react/src/hooks/useSettings.ts

```diff
@@ -246,53 +253,67 @@ const useSettings = <T>(config: SettingsConfig<T>): UseSettingsReturn<T> => {
         }
       }
     },
-    [user?.id, config.storagePath, settings],
+    [config.storagePath, loadableUser],
   );
 
   const resetSettings = useCallback(
-    async (settingsArray?: string[]) => {
-      if (!settings) return;
+    (settingsArray?: string[]) => {
+      const user = Loadable.match(loadableUser.get(), {
+        Loaded: (cUser) => cUser,
+        NotLoaded: () => undefined,
+      });
+      if (!user) return;
 
       const array = settingsArray ?? Object.keys(config.settings);
-      const newSettings = { ...settings };
 
-      array.forEach((setting) => {
-        let defaultSetting: SettingsConfigProp<T[Extract<keyof T, string>]> | undefined = undefined;
+      stateOb.update((s) => {
+        const news = s.get(config.storagePath);
+        array.forEach((setting) => {
+          let defaultSetting: SettingsConfigProp<T[Extract<keyof T, string>]> | undefined =
+            undefined;
 
-        for (const key in config.settings) {
-          const conf = config.settings[key];
+          for (const key in config.settings) {
+            const conf = config.settings[key];
 
-          if (conf.storageKey === setting) {
-            defaultSetting = conf;
-            break;
+            if (conf.storageKey === setting) {
+              defaultSetting = conf;
+              break;
+            }
           }
-        }
 
-        if (!defaultSetting) return;
+          if (!defaultSetting || !news) return;
 
-        newSettings[setting as keyof T] = defaultSetting.defaultValue;
+          news[setting] = defaultSetting.defaultValue;
+        });
+        return s.set(config.storagePath, news ?? {});
       });
-
-      update(config.storagePath, newSettings);
-
-      await updateDB(newSettings);
-
-      navigate('', { replace: true });
     },
-    [config, update, updateDB, navigate, settings],
+    [config, stateOb, loadableUser],
   );
 
   const updateSettings = useCallback(
-    async (updates: Settings, shouldPush = false) => {
-      if (!settings) return;
-
-      const newSettings = { ...settings, ...updates };
-
-      if (isEqual(newSettings, settings)) return;
+    (updates: Settings) => {
+      stateOb.update((s) =>
+        s.set(
+          config.storagePath,
+          s.get(config.storagePath) === updates
+            ? s.get(config.storagePath) ?? {}
+            : { ...s.get(config.storagePath), ...updates },
+        ),
+      );
+    },
+    [config, stateOb],
+  );
 
-      update(config.storagePath, newSettings);
+  useLayoutEffect(() => {
+    return derivedOb.subscribe(async (cur, prev) => {
+      const user = Loadable.match(loadableUser.get(), {
+        Loaded: (cUser) => cUser,
+        NotLoaded: () => undefined,
+      });
+      if (!cur || !user || cur === prev) return;
 
-      await updateDB(newSettings);
+      if (!initialLoading) await updateDB(cur, prev as SettingsRecord<T>);
```

Comment:
Another option here: since we're already setting up a subscription to `derivedOb` on [line 146](https://github.com/determined-ai/determined/pull/6202/files#diff-0ac40df189229b81aa5948c8d37e1c96bea9e253bdf2b233fef43661829242f3R146). You could use `usePrevious` to get access to the previous version and do this as a normal react `useEffect` but you'll need to be careful to avoid double-effects due to a dependency on both the current and previous versions in the `useEffect`.